hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/DotNode.java

```diff
@@ -583,39 +577,44 @@ public FromElement getImpliedJoin() {
 	}
 
 	/**
-	 * Is the given property name a reference to the primary key of the associated
-	 * entity construed by the given entity type?
+	 * Is the given property name a reference to the join key of the associated
+	 * entity constructed by the given entity type?
+	 * <p/>
+	 *
+	 * This method resolves the {@code propertyName} as an property of the entity type at the
+	 * {@link #propertyPath} relative to the {@link #getFromElement() FromElement}.
+	 * The implementation does so by invoking {@link EntityPersister#getPropertyType(String)},
+	 * which will resolve the property path against the entitiy's {@link org.hibernate.persister.entity.PropertyMapping}.
+	 * On initialisation of the {@link EntityPersister}, this {@code PropertyMapping} is filled with
+	 * property paths for all the owned properties and associations, and (embedded) identifier or unique key
+	 * properties for owned associations.
+	 * Henceforth, whenever a property path is found in the {@code PropertyMapping} of the {@code EntityPersister}
+	 * of the {@code FromElement}, we know that the property corresponds to a SQL fragment producible from the
+	 * {@code FromElement}, and as such the entity property can be dereferenced (optimised) in the final query.
+	 *
 	 * <p/>
 	 * For example, consider a fragment like order.customer.id
 	 * (where order is a from-element alias).  Here, we'd have:
 	 * propertyName = "id" AND
-	 * owningType = ManyToOneType(Customer)
-	 * and are being asked to determine whether "customer.id" is a reference
-	 * to customer's PK...
+	 * propertyPath = "customer"
+	 * FromElement = Order
+	 * and are being asked to determine whether "customer.id" is a property path of Order
 	 *
 	 * @param propertyName The name of the property to check.
-	 * @param owningType The type represeting the entity "owning" the property
 	 *
 	 * @return True if propertyName references the entity's (owningType->associatedEntity)
-	 *         primary key; false otherwise.
+	 *         join key; false otherwise.
 	 */
-	private boolean isReferenceToPrimaryKey(String propertyName, EntityType owningType) {
-		EntityPersister persister = getSessionFactoryHelper()
-				.getFactory()
-				.getEntityPersister( owningType.getAssociatedEntityName() );
-		if ( persister.getEntityMetamodel().hasNonIdentifierPropertyNamedId() ) {
-			// only the identifier property field name can be a reference to the associated entity's PK...
-			return propertyName.equals( persister.getIdentifierPropertyName() ) && owningType.isReferenceToPrimaryKey();
+	private boolean isPropertyEmbeddedInJoinProperties(String propertyName) {
+		EntityPersister entityPersister = this.getFromElement().getEntityPersister();
+		String propertyPath = String.join( ".", this.propertyPath, propertyName );
+		try {
+			Type propertyType = entityPersister.getPropertyType( propertyPath );
+			return propertyType != null;
```

Comment:
Are we really cautious enough? I might have missed something because I don't know this part of the code at all but I find it a bit weird that we only check that the entity has a property with the same name.

You could have a property with the same name that is not at all the foreign key for the other table and has a totally different value. How does the code work in this case?

Reply:
When we're in this method, we're wondering whether `someEntity.someAssociation.prop` can be derived from the SQL columns of `someEntity`. This method is only invoked when `someAssociation` refers to an `EntityType` (i.e. for embedded component types we already know that we don't have to join). The code related to this is in `DotNode#resolve(boolean, boolean, String, AST, AST)`.

The only "extended" paths (paths with a dot in it) that are added to the property mapping are either:
a. Embedded component types (`someEntity.someEmbeddable.someProp`). Obviously a join is not required here, but as mentioned before, this is already accounted for earlier in the process.
b. Referenced id or unique properties of associations (i.e. `someEntity.someAssociation.id` , or in this case `someEntity.someAssociation.isbn`).

Besides these, the property mapping also contains all basic properties. However, because we are in a `DotNode`, the `propertyPath` will always contain at least one dot, so we only have to consider the above two options here.

Furthermore, the exact same `PropertyMapping` is later used when actually dereferencing the attribute. (In `DotNode#getColumns()` > `FromElement#toColumns(String, String, boolean)` > `FromElementType.toColumns()`).

I'm pretty sure we're cautious enough here.